### ISSUE TITLE
List.triples duplicated

### ISSUE THREAD

**Issue by elkins0** [CindyJS/CindyJS#88]
This is just so that I don't forget to report things I find. This is large enough that you might care. You need to let me know to down what level of detail you want me to report. I'll add a comment pertaining to examples of some other things I've noticed.

List.triples appears twice in https://github.com/CindyJS/CindyJS/tree/master/src/js/libcs/List.js:
```javascript
List.triples = function(a) {
    var erg = [];
    for (var i = 0; i < a.value.length - 2; i++) {
        for (var j = i + 1; j < a.value.length - 1; j++) {
            for (var k = j + 1; k < a.value.length; k++) {
                erg.push({
                    'ctype': 'list',
                    'value': [a.value[i], a.value[j], a.value[k]]
                });
            }
        }
    }
    return {
        'ctype': 'list',
        'value': erg
    };
};

List.triples = function(a) {
    var erg = [];
    for (var i = 0; i < a.value.length - 2; i++) {
        for (var j = i + 1; j < a.value.length - 1; j++) {
            for (var k = j + 1; k < a.value.length; k++) {
                erg.push({
                    'ctype': 'list',
                    'value': [a.value[i], a.value[j], a.value[k]]
                });
            }
        }
    }
    return {
        'ctype': 'list',
        'value': erg
    };
};
```

**Comment by elkins0** [comment]
Here are examples of some other things I've noticed.

For instance, in https://github.com/CindyJS/CindyJS/tree/master/src/js/libcs/CSNumber.js we have:

```javascript
CSNumber.zero = CSNumber.real(0);
```

and yet there are 23 occurances of `CSNumber.real(0)` in the code.

In https://github.com/CindyJS/CindyJS/tree/master/src/js/libgeo/GeoOps.js,

```javascript
geoOps.Through.getParamForInput = function(el, pos, type) {
...
        l = List.turnIntoCSList([CSNumber.zero, CSNumber.zero, CSNumber.zero]);
```

using `l = List.realVector([0, 0, 0])` matches the more common way of doing this.

And here ...

```javascript
geoOps._helper.initializePoint = function(el) {
    var sx = 0;
    var sy = 0;
    var sz = 0;
    if (el.pos) {
        if (el.pos.length === 2) {
            sx = el.pos[0];
            sy = el.pos[1];
            sz = 1;
        }
        if (el.pos.length === 3) {
            sx = el.pos[0];
            sy = el.pos[1];
            sz = el.pos[2];
        }
    }
    var pos = List.realVector([sx, sy, sz]);
    pos = List.normalizeMax(pos);
    return pos;
};
```

where I would add an else and have `} else if (el.pos.length === 3) {`

**Comment by gagern** [comment]
> You need to let me know to down what level of detail you want me to report.

We know there are quite a number of features missing with regard to Cinderella compatibility. So as long as some deficiency is fairly obvious to spot, you may assume that we will take care of it once we got [more urgent stuff](https://github.com/CindyJS/CindyJS/labels/Cinderella%20compat.) dealt with. If you have code to contribute, that is of course welcome. If it's something subtle which might be missed, it's worth a report I think. Apart from that, stuff misbehaving is worth a bug report in any case. Bad coding style in the sources would be better addressed by a pull request which actually fixes the things in question. And in any case, it's important to have distinct reports for distinct issues, so we can prioritize and resolve them independently.

`List.triples` is indeed duplicate, and apparently has been since its introduction in a1ef53ba91e39125def81c14aa2374594b152a00. Please post this as a pull request, e.g. using the GitHub web interface.

Many of the `CSNumber.real(0)` occurrences predate the introduction of `CSNumber.zero`. I intend to clean them all up one day, and also have a constant for 1, and boolean constants for true and false. Haven't gotten round to this yet, and since it will affect many areas of code, it's a change which will likely conflict with other work, so I'd rather see some other work landed first. Keeping track of this in some ticket makes sense, but it should be separate from the one for `List.triples`.

The other two issues you mention would be nothing I'd consider worth fixing. `List.turnIntoCSList` causes less overhead than `List.realVector`, and re-using the `CSNumber.zero` constant reduces memory requirements, which can be important when considering small caches inside the CPU. So neither approach is clearly correct, both have merits. Likewise, not having an `else` makes things look more symmetric, and allows re-ordering the phrases, or formulating the first case in a way which uses the second as a fall-back (e.g. doing `el.pos.push(1)` or similar). Again both approaches have merits.

**Comment by gagern** [comment]
Main issue closed via #91. Feel free to post a separate issue about the use of constants like `CSNumber.zero` if you want to.
